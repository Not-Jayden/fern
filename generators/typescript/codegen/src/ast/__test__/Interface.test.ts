import { ts } from "../..";

describe("Interface", () => {
    describe("toString", () => {
        it("should generate a simple interface", () => {
            const interface_ = ts.interface_({
                name: "SimpleInterface",
                properties: []
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an exported interface", () => {
            const interface_ = ts.interface_({
                name: "ExportedInterface",
                properties: [],
                export: true
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an interface with one property", () => {
            const interface_ = ts.interface_({
                name: "OnePropertyInterface",
                properties: [{ name: "prop1", type: ts.Type.string() }]
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an interface with multiple properties", () => {
            const interface_ = ts.interface_({
                name: "MultiPropertyInterface",
                properties: [
                    { name: "prop1", type: ts.Type.string() },
                    { name: "prop2", type: ts.Type.number() },
                    { name: "prop3", type: ts.Type.boolean() }
                ]
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an interface with optional properties", () => {
            const interface_ = ts.interface_({
                name: "OptionalPropsInterface",
                properties: [
                    { name: "required", type: ts.Type.string() },
                    { name: "optional", type: ts.Type.number(), questionMark: true }
                ]
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an interface with readonly properties", () => {
            const interface_ = ts.interface_({
                name: "ReadonlyPropsInterface",
                properties: [
                    { name: "mutable", type: ts.Type.string() },
                    { name: "immutable", type: ts.Type.number() }
                ]
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an interface with complex types", () => {
            const interface_ = ts.interface_({
                name: "ComplexTypesInterface",
                properties: [
                    { name: "array", type: ts.Type.array(ts.Type.string()) },
                    { name: "literal", type: ts.Type.literal("literal-value") }
                ]
            });
            expect(interface_.toStringFormatted()).toMatchSnapshot();
        });

        it("should generate an interface extending another interface", () => {
            const extendedInterface = ts.interface_({
                name: "ExtendedInterface",
                properties: [{ name: "extendedProp", type: ts.Type.number() }],
                extends: [
                    ts.Reference.module({
                        module: "Sdk",
                        name: ["BaseSchema"],
                        source: "../../"
                    })
                ]
            });
            expect(extendedInterface.toString()).toMatchSnapshot();
        });
    });
});
