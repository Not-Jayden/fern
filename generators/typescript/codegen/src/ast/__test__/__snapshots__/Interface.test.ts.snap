// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Interface > toString > should generate a simple interface 1`] = `
"interface SimpleInterface {}
"
`;

exports[`Interface > toString > should generate an exported interface 1`] = `
"export interface ExportedInterface {}
"
`;

exports[`Interface > toString > should generate an interface extending another interface 1`] = `
"import * as Sdk from "../.";

 interface ExtendedInterface extends Sdk.BaseSchema{
    "extendedProp": number;
}
"
`;

exports[`Interface > toString > should generate an interface with complex types 1`] = `
"interface ComplexTypesInterface {
    array: string[];
    literal: "literal-value";
}
"
`;

exports[`Interface > toString > should generate an interface with multiple properties 1`] = `
"interface MultiPropertyInterface {
    prop1: string;
    prop2: number;
    prop3: boolean;
}
"
`;

exports[`Interface > toString > should generate an interface with one property 1`] = `
"interface OnePropertyInterface {
    prop1: string;
}
"
`;

exports[`Interface > toString > should generate an interface with optional properties 1`] = `
"interface OptionalPropsInterface {
    required: string;
    optional?: number;
}
"
`;

exports[`Interface > toString > should generate an interface with readonly properties 1`] = `
"interface ReadonlyPropsInterface {
    mutable: string;
    immutable: number;
}
"
`;
